dvc/pkg/base.py

```diff
@@ -0,0 +1,94 @@
+from __future__ import unicode_literals
+
+import os
+import copy
+import shutil
+
+from dvc.utils.compat import urlparse, str, builtin_str
+
+from dvc.config import Config
+from dvc.remote.config import remote_add
+
+from .config import get_pkg_settings
+
+
+class PkgBASE(object):
+    PKG_DIR = "pkg"
+
+    def __init__(self, config, settings):
+        self.settings = copy.copy(settings)
+
+        name = None
+        if isinstance(settings, str) or isinstance(settings, builtin_str):
+            name = settings
+            settings = {}
+        elif isinstance(settings, dict):
+            name = settings.pop("name")
+
+        if os.path.exists(name) or urlparse(name).scheme:
+            if Config.SECTION_PKG_URL not in settings:
+                settings[Config.SECTION_PKG_URL] = name
+            name = os.path.basename(name)
+            config_settings = {}
+        else:
+            config_settings = get_pkg_settings(config, name) if name else {}
+
+        config_settings.update(settings)
+
+        self.url = config_settings.get(Config.SECTION_PKG_URL)
+        self.revision = config_settings.get(Config.SECTION_PKG_REVISION)
+
+        self._dir = os.path.join(config.dvc_dir, self.PKG_DIR, name)
+
+        self._repo = None
+
+    @property
+    def repo(self):
+        if not self._repo or self._repo.root_dir != self._dir:
+            from dvc.repo import Repo
+
+            revision = self.revision
+            if revision:
+                revision = "remotes/origin/" + revision
+
+            self._repo = Repo(self._dir, revision)
+
+        return self._repo
+
+    @property
+    def installed(self):
+        return os.path.isdir(self._dir)
+
+    def _setup_remote(self):
+        core = self.repo.config.config.get(Config.SECTION_CORE, {})
+        remote = core.get(Config.SECTION_CORE_REMOTE)
+
+        if not remote and os.path.isdir(self.url):
+            from dvc.repo import Repo
+
+            repo = Repo(self.url)
```

Comment:
self.repo and repo - can we rename one of them - hard to read when the only difference is `self.`